### PR TITLE
MM-15010 Adding Cypress test for - Pinning or un-pinning older post d…

### PR DESCRIPTION
**Pull request by prapti**
#### Summary
Added E2E test using `cypress` for  "Pinning or un-pinning older post does not cause it to display at bottom of channel"

1. Added `message_pinning_unpinning_spec.js` for E2E test
2. Added `apiUnpinPosts()` method in `api_commands.js` to easily unpin posts at the end of test
3. Added `getNthPostId()` method in `ui_commands.js` to find and navigate nth number of post in the CENTER display

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10575
Jira: https://mattermost.atlassian.net/browse/MM-15010

### CODE REVIEW

**Comment by saturninoabril** [comment]
@prapti fyi, I'm offline today.  Will give a review of this first thing tomorrow.

**Comment by prapti** [comment]
No problem, thank you @saturninoabril

File: cypress/integration/onboarding/messaging/message_pinning_unpinning_spec.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    // Unpin all posts at the end of the test
+    after(() => {
+        pinnedPosts.forEach((pinnedPost) => {
+            cy.apiUnpinPosts(pinnedPost);
+        });
+    });
+
+    it('M15010 Pinning or un-pinning older post does not cause it to display at bottom of channel', () => {
+        cy.get('#sidebarItem_saepe-5').click();
```
Comment (saturninoabril) [inline cypress/integration/onboarding/messaging/message_pinning_unpinning_spec.js]: For some reason, this is failing on me.  One way is to do force click [`({force: true})`]

File: cypress/integration/onboarding/messaging/message_pinning_unpinning_spec.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                    // * Right-hand-side only has 2 initially pinned posts
+                    cy.get('#search-items-container').children().should('have.length', 2);
+
+                    // * Right-hand-side does not have the post pinned in step 4.
+                    cy.get('#search-items-container').children().should('not.contain', `#postMessageText_${postId}`);
+                });
+            });
+        });
+    });
+});
```
Comment (saturninoabril) [inline cypress/integration/onboarding/messaging/message_pinning_unpinning_spec.js]: 💯

File: cypress/support/api_commands.js [hunk 3]
```diff
@@ -233,3 +233,20 @@ Cypress.Commands.add('apiEnableOpenServer', (enable = true) => {
...
+* This API assume that the user is logged in and has cookie to access
+* @param {Object} postId - Post ID of the pinned post to unpin
+*/
+Cypress.Commands.add('apiUnpinPosts', (postId) => {
+    return cy.request({
+        headers: {'X-Requested-With': 'XMLHttpRequest'},
+        url: '/api/v4/posts/' + postId + '/unpin',
+        method: 'POST',
+    });
+});
```
Comment (saturninoabril) [inline cypress/support/api_commands.js]: 💯

File: cypress/support/ui_commands.js [hunk 4]
```diff
@@ -233,6 +233,17 @@ Cypress.Commands.add('getLastPostIdWithRetry', () => {
...
+* Get post ID for nth newest post
+* .eq() is 0-based index, hence nthPost-2 to get the nth post
+@param {Integer} nthPost - nth newest post
+*/
+Cypress.Commands.add('getNthPostId', (nthPost) => {
+    return cy.get('#postListContent [id^=post]:first').parent().parent().siblings().eq(nthPost - 2).find('[id^=post]:first').invoke('attr', 'id').then((nthPostId) => {
+        return nthPostId.replace('post_', '');
+    });
+});
+
```
Comment (saturninoabril) [inline cypress/support/ui_commands.js]: 💯

File: cypress/integration/onboarding/messaging/message_pinning_unpinning_spec.js (hunk 1 above)
Comment (prapti) [reply cypress/integration/onboarding/messaging/message_pinning_unpinning_spec.js]: Added the forced click there. Hopefully it'll run now :)

File: cypress/support/ui_commands.js (hunk 4 above)
Comment (cometkim) [reply cypress/support/ui_commands.js]: I'm little confused on selector here, I got null when I tried to validate this on community.mattermost.com

@prapti Could you help me to understand this?

image

**Comment by hanzei** [comment]
Hey @cometkim, Gently reminder about reviewing this PR :wink:

File: cypress/support/ui_commands.js (hunk 4 above)
Comment (prapti) [reply cypress/support/ui_commands.js]: @cometkim This test was developed and verified against test-data and not necessarily against  community server. Is that a common approach for Cypress tests? But the general idea behind the step was to find out the older posts in the CENTER layout with a given post ID, similar to how the last post is identified.

**Pull request merged**